primitives/trie/src/trie.rs

```diff
@@ -469,13 +508,72 @@ impl MerkleRadixTrie {
+                | e @ Err(MerkleRadixTrieError::ChildDoesNotExist) => {
                     // We cannot continue further down as the next node is outside our trie.
                     // Update this node prior to `update_keys` (which only updates the other nodes in the root path).
                     node.put_child_no_hash(key).expect("Prefix must be correct");
                     self.put_node(txn, &node, OldValue::Unchanged);
 
+                    // Mark that a new stump has been added.
+                    if let Err(MerkleRadixTrieError::ChildDoesNotExist) = e {
+                        new_stump = true;
+                    }
+
+                    root_path.push(node);
+                    break;
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+        }
+
+        self.update_keys(txn, root_path, CountUpdates::default());
+
+        Ok(new_stump)
+    }
+
+    /// Removes the corresponding stump and updates the hashes for a path
+    /// to a node in the missing part of the tree.
+    /// This is used to indicate that the corresponding branch's hash has changed,
+    /// but we cannot provide the accurate hash until a new chunk has been pushed.
+    fn remove_within_missing_part_raw(
+        &self,
+        txn: &mut WriteTransactionProxy,
+        key: &KeyNibbles,
+        missing_range: &Option<ops::RangeFrom<KeyNibbles>>,
+    ) -> Result<(), MerkleRadixTrieError> {
+        let mut node = self
+            .get_root(txn)
+            .expect("The Merkle Radix Trie didn't have a root node!");
+
+        let mut root_path = Vec::new();
+        // Descend down the tree and collect nodes to be updated.
+        loop {
+            // This function should only be called with keys in the missing range.
+            assert_ne!(&node.key, key);
+
+            // Check that the key we are trying to update can exist in this part of the tree.
+            if !node.key.is_prefix_of(key) {
+                return Err(MerkleRadixTrieError::ChildDoesNotExist);
+            }
+
+            // Descend to the corresponding child.
+            match node.child_key(key, missing_range) {
+                Ok(child_key) => {
+                    let child = self
+                        .get_node(txn, &child_key)
+                        .expect("Child should be present");
+                    let parent = mem::replace(&mut node, child);
+                    root_path.push(parent);
+                }
+                Err(MerkleRadixTrieError::ChildIsStump) => {
+                    // We cannot continue further down as the next node is outside our trie.
+                    // Remove this stump.
+                    node.remove_child(key).expect("Prefix must be correct");
+                    self.put_node(txn, &node, OldValue::Unchanged);
+
                     root_path.push(node);
                     break;
                 }
```

Comment:
Is `Err(MerkleRadixTrieError::ChildDoesNotExist` here `unreachable!`? If so I would prefer it to panic if encountered.